### PR TITLE
Copy data from old map to the new map when config changes

### PR DESCRIPTION
**Pull request by sridhartigera**
## Description
When creating the map with new size, copy data from the old map to the new map so that we so not lose data and thus disturb operation.

Some of the maps like conntrack are being updated by BPF programs while userspace is migrating the data. That means a single sweep of the old map may not copy all the data. Therefore we do another sweep once we switch the BPF programs from using the old map to the new ones and copy the data that we missed in the first sweep when copying from the old map. This reduces the chance of disruption.

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

**Review by fasaxc — CHANGES_REQUESTED** [review]

File: felix/bpf/maps.go [hunk 1]
```diff
@@ -299,6 +305,43 @@ func (b *PinnedMap) Delete(k []byte) error {
...
+		if numEntriesCopied == b.MaxEntries {
+			return fmt.Errorf("new map cannot hold all the data from the old map.")
+		}
+
+		err = b.Update(k, v)
+		if err != nil {
+			return fmt.Errorf("error copying data from the old map")
+		}
+		logrus.Debugf("copied data from old map to new map key=%v, value=%v", k, v)
+		numEntriesCopied++
```
[the remark this answers is not in the document]
Comment (tomastigera) [reply felix/bpf/maps.go:344]: Yes, that could happen for multiple reasons. We addressed some of it in the iterator itself, didn't we? It was a while back :thinking:

File: felix/bpf/maps.go [hunk 2]
```diff
@@ -299,6 +302,47 @@ func (b *PinnedMap) Delete(k []byte) error {
...
+		if err != nil {
+			if err == ErrIterationFinished {
+				return nil
+			}
+			return errors.Errorf("iterating the old map failed: %s", err)
+		}
+
+		if numEntriesCopied == b.MaxEntries {
+			return fmt.Errorf("new map cannot hold all the data from the old map.")
+		}
```
Comment (tomastigera) [inline felix/bpf/maps.go:331]: What if you just wait for E2BIG error (iirc that is the error returned when the map is full) I wonder if this is always an error. for instance LRU map is likely going to be full but the entries may be expired (say affinity map), perhaps we need the map instance to decide whether the entry needs to be copied. Then you could downsize a full map. Maybe something to note for the next iteration?

File: felix/bpf/maps.go [hunk 3]
```diff
@@ -397,12 +444,17 @@ func (b *PinnedMap) EnsureExists() error {
...
+		b.oldSize = mapInfo.MaxEntries
+
+		err = b.repinAt(b.Path(), oldMapPath)
 		if err != nil {
 			return fmt.Errorf("error migrating the old map %w", err)
 		}
+		copyData = true
+		defer func() {
+			b.oldfd.Close()
+			b.oldfd = 0
```
Comment (tomastigera) [inline felix/bpf/maps.go:456]: i think -1 would be a better value as 0 is a valid value

File: felix/fv/bpf_map_resize_test.go [hunk 4]
```diff
@@ -84,6 +88,35 @@ var _ = infrastructure.DatastoreDescribe("_BPF-SAFE_ Felix bpf test configurable
 		}
 	}
 
+	It("should copy data from old map to new map", func() {
```
Comment (tomastigera) [inline felix/fv/bpf_map_resize_test.go:91]: It feels like some UTs in bpf package might be more light-way easier to control tests for this. That would allow you to more easily test the content (also if expired values are being removed along the way), all the corner cases, error cases etc.

File: felix/bpf/ut/maps_test.go [hunk 5]
```diff
@@ -26,9 +27,167 @@ import (
...
+
+	// Total number of k,v in old map should be 1
+	seenKeys := map[conntrack.Key]bool{}
+	err = mc.CtMap.Iter(func(k, v []byte) bpf.IteratorAction {
+		key := conntrack.KeyFromBytes(k)
+		Expect(seenKeys[key]).To(BeFalse(), "Saw a duplicate key")
+		seenKeys[key] = true
+		return bpf.IterNone
+	})
+	Expect(err).NotTo(HaveOccurred())
```
Comment (tomastigera) [inline felix/bpf/ut/maps_test.go:163]: You can do `ctSaved := saveCTMap(mc.ctMap)`

Are you going to se a duplicate key here? I do not think so, nobody is updating the map simultaneously.

File: felix/bpf/ut/maps_test.go [hunk 6]
```diff
@@ -26,9 +27,167 @@ import (
...
+	mc := bpfmap.CreateBPFMapContext(ipsetsMapSize, natFeMapSize, natBeMapSize, natAffMapSize, rtMapSize, ctMapSize, true)
+	bpfmap.CreateBPFMaps(mc)
+	defer func() {
+		restoreMaps(mc)
+	}()
+	// New CT map should have max_entries as 600
+	ctMapInfo, err := bpf.GetMapInfo(mc.CtMap.MapFD())
+	Expect(err).NotTo(HaveOccurred(), "Failed to get ct map info")
+	Expect(ctMapInfo.MaxEntries).To(Equal(ctMapSize))
+	// Except CT map, other map's old pins should be deleted.
```
[the remark this answers is not in the document]
Comment (sridhartigera) [reply felix/bpf/ut/maps_test.go:67]: yeah. We dont delete the old map for conntrack as we defer the copy.

**Review by tomastigera — APPROVED** [review]
LGTM, thanks
